docs/proposals/image-index.md

```diff
@@ -0,0 +1,150 @@
+based on *base* image).
+
+## Use Cases
+
+* Classify a certain image by a label and have that label propagated to all descentants of that
+  image currently present in the cluster.
+  * By classification, we are giving an Image context, or role in the cluster. For example an image
+    classified as "VENDOR" represents all images that are based on this "VENDOR" image.
+* Be able to query:
+  * ImageStreamTags that currently reference an Image classified as *foo*
+  * ImageStreamTags that reference outdated version of an Image classified as *foo*
+  * All Images based on the Image *foo* (descentants)
+  * All Images that the *foo* images is based on (ancestors)
+* Be able to track the usage of images limited by an license by giving the image carrying the licenced
+  content a class
+* Be able to enforce policies based on classification:
+  * Allow running only *VENDOR* based images in this cluster
+* Be able to perform bulk operations based on the classification:
+  * Trigger a build for all BuildConfigs that have builder image based on *VENDOR*.
+
+### Current State
+
+The information about the Image(s) we gather is not normalized in any way. It
+does not provide any information about relationships between images, other than
+their affiliation with ImageStream. Furthermore, all of the information
+currently available is scattered throughout the entire system, and accessed
+from different parts of the system in non-unified way.
+
+This results in the tools such as pruning or `oadm top images` always build
+in-memory state of the cluster for their operation.
+
+### Goal
+
+To build efficient *Image Index API* that allows to querying the relations
+between images and also name and map these relations into human-readable form.
+
+### Known Limitations / Scope Of Work
+
+* We can only keep track of Images we know about (iow. imported into OpenShift)
+  * If somebody use an external image, we can’t map it to indexing API without importing
+    it into an Image object
+* Currently, there is no way to restrict users referencing an external images in Pods.
+  * Solving this problem goes beyond the scope of this proposal
+* This API is meant to only query the images/imagestreams not their usage in the Pods.
+* The classification is meant to be managed and used by cluster/registry administrators,
+  not by regular users
+
+## Proposed Workflow
+
+* As an admin I want to identify all *VENDOR* based images in the cluster:
+  * I create new `classification` using the *VENDOR* ImageStream(Tag)
+  * `ImageClassificationController` discovers all descentants of this Image (or any previous
+     version of this Image) in the cluster
+  * `ImageClassificationController` applies label/annotations to the Image(s) that include
+    reference to *VENDOR*
+
+* As an admin I want to group the classifications together, so I can send "key=value" queries, for example:
+  * I want to get all images with `os=rhel7` or `vendor=redhat` or `lang=python`:
```

Comment:
The question above relates to human querying, not layer inferencing (which needs the efficient query for layer membership).  But layer membership isn't enough - pulling an image to a docker engine compiled in Go 1.6 and pushing back to a 1.7 registry could change that (because of tar archive stuff).  So any sort of "scanning engine" has to be more complicated than just what we provide.